src/addr_families.c

```diff
@@ -0,0 +1,832 @@
+
+    if ((ret = wolfsentry_table_ent_delete_1(wolfsentry, &old->header)) < 0)
+        return ret;
+#ifdef WOLFSENTRY_PROTOCOL_NAMES
+    if ((ret = wolfsentry_table_ent_delete_1(wolfsentry, &old->byname_ent->header)) < 0)
+        return ret;
+#endif
+
+    return wolfsentry_addr_family_drop_reference(wolfsentry, old, action_results);
+}
+#endif
+
+wolfsentry_errcode_t wolfsentry_addr_family_handler_install(
+    struct wolfsentry_context *wolfsentry,
+    wolfsentry_addr_family_t family_bynumber,
+    const char *family_byname, /* if defined(WOLFSENTRY_PROTOCOL_NAMES), must not NULL, else ignored. */
+    int family_byname_len,
+    wolfsentry_addr_family_parser_t parser,
+    wolfsentry_addr_family_formatter_t formatter,
+    int max_addr_bits)
+{
+    return wolfsentry_addr_family_insert(wolfsentry, wolfsentry->addr_families_bynumber, family_bynumber, family_byname, family_byname_len, parser, formatter, max_addr_bits);
+}
+
+wolfsentry_errcode_t wolfsentry_addr_family_handler_remove_bynumber(
+    struct wolfsentry_context *wolfsentry,
+    wolfsentry_addr_family_t family_bynumber,
+    wolfsentry_action_res_t *action_results)
+{
+    return wolfsentry_addr_family_handler_delete_bynumber(wolfsentry, wolfsentry->addr_families_bynumber, family_bynumber, action_results);
+}
+
+#ifdef WOLFSENTRY_PROTOCOL_NAMES
+wolfsentry_errcode_t wolfsentry_addr_family_handler_remove_byname(
+    struct wolfsentry_context *wolfsentry,
+    const char *family_byname,
+    int family_byname_len,
+    wolfsentry_action_res_t *action_results)
+{
+    return wolfsentry_addr_family_handler_delete_byname(wolfsentry, wolfsentry->addr_families_byname, family_byname, family_byname_len, action_results);
+}
+#endif
+
+#ifdef WOLFSENTRY_PROTOCOL_NAMES
+
+wolfsentry_addr_family_t wolfsentry_addr_family_pton(
+    struct wolfsentry_context *wolfsentry,
+    const char *family_name,
+    int family_name_len,
+    wolfsentry_errcode_t *errcode)
+{
+    wolfsentry_errcode_t ret;
+
+    if (family_name == NULL) {
+        if (errcode != NULL)
+            *errcode = WOLFSENTRY_ERROR_ENCODE(INVALID_ARG);
+        return WOLFSENTRY_AF_UNSPEC;
+    }
+
+    if (family_name_len < 0)
+        family_name_len = (int)strlen(family_name);
+
+    {
+        struct wolfsentry_addr_family_bynumber *addr_family;
+        ret = wolfsentry_addr_family_get_byname_1(
+            wolfsentry->addr_families_byname,
+            family_name,
+            (int)family_name_len,
+            &addr_family);
+        if (WOLFSENTRY_ERROR_CODE_IS(ret, OK)) {
+            if (errcode)
+                *errcode = ret;
+            return addr_family->number;
+        } else if (! WOLFSENTRY_ERROR_CODE_IS(ret, ITEM_NOT_FOUND)) {
+            if (errcode != NULL)
+                *errcode = ret;
+            return WOLFSENTRY_AF_UNSPEC;
+        }
+    }
+
+    if (errcode)
+        *errcode = WOLFSENTRY_ERROR_ENCODE(OK);
+    if (strcaseeq(family_name, "UNIX", (size_t)family_name_len))
```

Comment:
Would this section be better as a lookup table?

Reply:
doubtless better as a lookup table.  let's leave it lame for now (avoid premature optimization) and promise we'll fix it later.

the biggest yet-unseized optimization opportunity is red-black tables for the main tables.  it's on the roadmap.